api/include/opentelemetry/metrics/instrument.h

```diff
@@ -0,0 +1,191 @@
+#pragma once
+
+#include "opentelemetry/common/attribute_value.h"
+#include "opentelemetry/nostd/shared_ptr.h"
+#include "opentelemetry/nostd/string_view.h"
+#include "opentelemetry/trace/key_value_iterable_view.h"
+
+OPENTELEMETRY_BEGIN_NAMESPACE
+namespace metrics
+{
+
+// Enum classes to help determine instrument types in other parts of the API
+enum class InstrumentKind
+{
+  IntCounter,
+  IntUpDownCounter,
+  IntValueRecorder,
+  IntValueObserver,
+  IntSumObserver,
+  IntUpDownSumObserver,
+  DoubleCounter,
+  DoubleUpDownCounter,
+  DoubleValueRecorder,
+  DoubleValueObserver,
+  DoubleSumObserver,
+  DoubleUpDownSumObserver
+};
+
+// Fewer Bound types because Asynchronous instruments cannot bind
+enum class BoundInstrumentKind
+{
+  BoundIntCounter,
+  BoundIntUpDownCounter,
+  BoundIntValueRecorder,
+  BoundDoubleCounter,
+  BoundDoubleUpDownCounter,
+  BoundDoubleValueRecorder
+};
+
+class Instrument
+{
+
+public:
+  // Note that Instruments should be created using the Meter class.
+  // Please refer to meter.h for documentation.
+  Instrument() = default;
+
+  /**
+   * Base class constructor for all other instrument types.  Whether or not
+   * an instrument is synchronous or bound, it requires a name, description,
+   * unit, and enabled flag.
+   *
+   * @param name is the identifier of the instrumenting library
+   * @param description explains what the metric captures
+   * @param unit specifies the data type held in the instrument
+   * @param enabled determines if the metric is currently capturing data
+   * @return Instrument type with the specified attirbutes
+   */
+  Instrument(nostd::string_view name,
+             nostd::string_view description,
+             nostd::string_view unit,
+             bool enabled)
+  {}
+
+  // Returns true if the instrument is enabled and collecting data
+  virtual bool IsEnabled()
+  {
+    // False in default implementation
+    return false;
+  }
+
+  // Return the instrument name
+  virtual nostd::string_view GetName() { return nostd::string_view(""); }
```

Comment:
Do we want the default implementation or this could be pure virtual?

Reply:
I'll make that change, though the tradeoff is substantial code duplication in the noop instruments further down the road.